### DISCUSSION TITLE
Drop Python 3.6 support

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#2005] [General]
I don't want to drop support for a Python version without good reasons, but I also don't want to keep the support / increase the inconvenience to extremely.

This discussion will remain open until the reasons to drop 3.6 become stronger than the reasons to keep the support.

If you have an opinion on the topic, you can vote on one of the two comments of this post: drop support / keep support.

edit: I'm trying to get some data on [Which Python version are you using?](https://twitter.com/py_pdf/status/1685171134789001216)

edit: Done via https://github.com/py-pdf/pypdf/pull/2369

**Comment by MartinThoma** [comment]
Drop support reasons:

* [Python 3.6 reached it's end-of-live](https://peps.python.org/pep-0494/) in December 2021. Even [Python 3.7 is no longer supported](https://endoflife.date/python)
* [Killer features of newer Python versions](https://python.plainenglish.io/killer-features-by-python-version-c84ca12dba8):
    * [Python 3.7 has dataclasses](https://peps.python.org/pep-0557/)
* [pypdf downloads](https://pypistats.org/packages/pypdf): Very few Python 3.6 downloads as of July 2023
  * 57.49% use Python 3.8+
  * 9.68% NULL (no data)
  * 6.85% Python 3.7
  * 0.33% Python 3.6
  * 1.18% Python 2.7 :exploding_head: 
* Important packages dropped support:
    * [Cryptography](https://pypi.org/project/cryptography/) is Python 3.7+. Other projects like Pycryptodome might follow.
    * [mypy==0.981](https://mypy-lang.blogspot.com/2022/09/mypy-0981-released.html): September 2022
    * [typing_extensions==4.2.0](https://github.com/python/typing_extensions/blob/main/CHANGELOG.md#release-420-april-17-2022): April 2022
    * [Pillow==9.0.0](https://pillow.readthedocs.io/en/stable/releasenotes/9.0.0.html): January 2022 - there are several critical issues that require 9.0.1, e.g. https://github.com/advisories/GHSA-pw3c-h7wp-cvhx
* Comparable packages:
    * [PyMuPDF==1.23.0](https://pymupdf.readthedocs.io/en/latest/changes.html) dropped support for Python 3.7 in August 2023
    * [pikepdf==v7.2.0](https://pikepdf.readthedocs.io/en/latest/releasenotes/version7.html) dropped support for Python 3.7 in April 2023
    * [fpdf2==2.7.0](https://pypi.org/project/fpdf2/) released in March 2023 also requires Python 3.7.
* OS default versions:
   * [Ubuntu 20.04](https://askubuntu.com/a/1232820) has 3.8.2. That is the [oldes Ubuntu version](https://en.wikipedia.org/wiki/Ubuntu_version_history) still stupported
   * [Debian 10](https://linuxize.com/post/how-to-install-python-3-9-on-debian-10/) has Python 3.7. That is the [oldest Debian version still supported](https://en.wikipedia.org/wiki/Debian_version_history)
 * Services:
   * [AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtime-support-policy): August 2022
   * [Azure Functions](https://azure.microsoft.com/en-us/updates/azure-functions-support-for-python-36-is-ending-on-30-september-2022/): September 2022
   * [Heroku](https://devcenter.heroku.com/changelog-items/2546): February 2023

**Comment by MartinThoma** [comment]
Keep Python 3.6 support:

* [Web servers still use 3.6 more often than any other version](https://w3techs.com/technologies/details/pl-python/3) (July 2023: 31.5%, 2023-10-01: 28.8%)
* OS default versions:
    * [RHEL 8 uses Python 3.6](https://developers.redhat.com/blog/2018/11/14/python-in-rhel-8-3#:~:text=In%20RHEL%208%2C%20Python%203.6,install%20python3%20to%20get%20it.) and is still active and has full support until 31. May 2024. The https://github.com/py-pdf/pypdf/discussions/683 discussion shows RHEL 7 was still used in 2022
    * [openSUSE Leap 15.4 ](https://en.opensuse.org/Lifetime): maintained until end of December 2024; system Python is 3.6, but 3.11 is available ([source](https://en.opensuse.org/Release_announcement_15.5))

**Reply by stefan6419846** [reply]
The current OpenSUSE Leap 15.5, supported until December 2024, ships with Python 3.6 by default (as does SLES/SUSE Linux Enterprise Server), although parallel installations of newer distributions are possible: https://en.opensuse.org/Features_15.5#Python
